pkg/stanza/fileconsumer/internal/reader/reader.go

```diff
@@ -106,9 +106,90 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 		}
 	}()
 
+	doneReadingFile := r.readHeader(ctx)
+	if doneReadingFile {
+		return
+	}
+
+	// Reset position in file to r.Offest after the header scanner might have moved it past a content token.
+	if _, err := r.file.Seek(r.Offset, 0); err != nil {
+		r.set.Logger.Error("failed to seek post-header", zap.Error(err))
+		return
+	}
+
+	r.readContents(ctx)
+}
+
+func (r *Reader) readHeader(ctx context.Context) (doneReadingFile bool) {
+	if r.headerReader == nil {
+		r.set.Logger.Debug("no need to read header", zap.Bool("header_finalized", r.HeaderFinalized))
+		return false
+	}
+
 	s := scanner.New(r, r.maxLogSize, r.initialBufferSize, r.Offset, r.splitFunc)
 
-	// Iterate over the tokenized file, emitting entries as we go
+	// Read the tokens from the file until no more header tokens are found or the end of file is reached.
+	for {
+		select {
+		case <-ctx.Done():
+			return true
+		default:
+		}
+
+		ok := s.Scan()
+		if !ok {
+			if err := s.Error(); err != nil {
+				r.set.Logger.Error("failed during header scan", zap.Error(err))
+			} else {
+				r.set.Logger.Debug("end of file reached", zap.Bool("delete_at_eof", r.deleteAtEOF))
+				if r.deleteAtEOF {
+					r.delete()
+				}
+			}
+			// Either end of file was reached, or file cannot be scanned.
+			return true
+		}
+
+		token, err := r.decoder.Decode(s.Bytes())
+		if err != nil {
+			r.set.Logger.Error("decode header: %w", zap.Error(err))
+			r.Offset = s.Pos() // move past the bad token or we may be stuck
+			continue
+		}
+
+		err = r.headerReader.Process(ctx, token, r.FileAttributes)
+		if err != nil {
+			if errors.Is(err, header.ErrEndOfHeader) {
+				// End of header reached.
+				break
+			} else {
+				r.set.Logger.Error("process header: %w", zap.Error(err))
+			}
+		}
+
+		r.Offset = s.Pos()
+	}
+
+	// Clean up the header machinery
+	if err := r.headerReader.Stop(); err != nil {
+		r.set.Logger.Error("failed to stop header pipeline during finalization", zap.Error(err))
+	}
+	r.headerReader = nil
+	r.HeaderFinalized = true
+	r.initialBufferSize = scanner.DefaultBufferSize
+
+	// Switch to the normal split and process functions.
+	r.splitFunc = r.lineSplitFunc
+	r.processFunc = r.emitFunc
```

Comment:
This "juggling" of function references could probably be avoided altogether after the refactoring.